File: app/src/main/java/com/decode/tumblr/posts/FragmentMain.java
package com.decode.tumblr.posts;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.decode.tumblr.R;
import com.decode.tumblr.model.PostObject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class FragmentMain extends Fragment implements RecycleViewAdapter.OnPostClickListener {

    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;
    @BindView(R.id.swipeRefreshLayout)
    SwipeRefreshLayout swipeRefreshLayout;
    @BindView(R.id.progress_bar)
    ProgressBar progressBar;

    private RecycleViewAdapter adapter;
    private PostViewModel postViewModel;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.layout_fragment_main, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        // Setup recyclerView
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setItemAnimator(new DefaultItemAnimator());

        // Setup adapter
        adapter = new RecycleViewAdapter();
        recyclerView.setAdapter(adapter);
        adapter.setOnPostClickListener(FragmentMain.this);

        // Get data
        postViewModel = ViewModelProviders.of(this).get(PostViewModel.class);

        subscribePostData();

        // On swipe - get new data from server
        swipeRefreshLayout.setOnRefreshListener(() -> {
            swipeRefreshLayout.setRefreshing(true);
            postViewModel.fetchPosts();
        });
    }

    @Override
    public void onPostClick(PostObject post, View view) {
        FragmentMainDirections.ActionFragmentMainToFragmentDetails action =
                FragmentMainDirections.actionFragmentMainToFragmentDetails(post);
        Navigation.findNavController(view).navigate(action);
    }

    private void subscribePostData() {
        postViewModel.getPosts().observe(this, posts -> {
            adapter.setPostList(posts);
            progressBar.setVisibility(View.GONE);
            swipeRefreshLayout.setRefreshing(false);
        });

        postViewModel.getLoadingFailed().observe(this, error ->
                Toast.makeText(getContext(), error, Toast.LENGTH_SHORT).show());
    }
}

File: app/src/main/java/com/decode/tumblr/posts/FragmentMain.kt
package com.decode.tumblr.posts

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.navigation.Navigation
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.decode.tumblr.R
import com.decode.tumblr.data.db.model.PostObject
import kotlinx.android.synthetic.main.layout_fragment_main.*
import org.koin.android.viewmodel.ext.android.viewModel

class FragmentMain : Fragment(), PostsAdapter.OnPostClickListener {

    private val postViewModel: PostViewModel by viewModel()
    private var adapter: PostsAdapter? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.layout_fragment_main, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        // Setup recyclerView
        recyclerView!!.setHasFixedSize(true)
        recyclerView!!.layoutManager = LinearLayoutManager(context)
        recyclerView!!.itemAnimator = DefaultItemAnimator()

        // Setup adapter
        adapter = PostsAdapter()
        recyclerView!!.adapter = adapter
        adapter!!.setOnPostClickListener(this@FragmentMain)

        subscribePostData()

        // On swipe - get new data from server
        swipeRefreshLayout!!.setOnRefreshListener {
            swipeRefreshLayout!!.isRefreshing = true
            postViewModel.fetchPosts()
        }
    }

    override fun onPostClick(post: PostObject, view: View) {
        val action = FragmentMainDirections.actionFragmentMainToFragmentDetails(post)
        Navigation.findNavController(view).navigate(action)
    }

    private fun subscribePostData() {
        postViewModel.posts.observe(this, Observer { posts ->
            adapter!!.setPostList(posts)
            progressBar!!.visibility = View.GONE
            swipeRefreshLayout!!.isRefreshing = false
        })

        postViewModel.loadingFailed.observe(this, Observer { error ->
            Toast.makeText(context, error, Toast.LENGTH_SHORT).show()
        })
    }
}
